### PR TITLE
Queues - Hyunji Kim - Scrabble

### PR DESCRIPTION
**Pull request by ricecakemonster**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   How do you feel you and your partner did in sharing responsibilities?	|   While we did it together(the first day), it was really great.	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |In that way, we can use the "score" method in other classes without making a instance of Scoring class.    |
|   Describe an example of a test you wrote for a _nominal case_. |  Player cannot keep on playing after they already have won. 	|
|   Describe an example of a test you wrote for an _edge case_. | When top score is tied between multiple words and one used all seven letters, return the one with seven letters  	|
|   What was your final test coverage percentage?	|   100 percent	|
|   Is there a specific piece of code you'd like feedback on?	|   no	|

### CODE REVIEW

File: lib/scrabble_scoring.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
+module Scrabble
+  class Scoring
+    LETTERS = {
```
Comment (droberts-sea) [inline lib/scrabble_scoring.rb:3]: This hash of arrays definitely works, but looking up a letter's score ends up being complicated, involving an `each` loop.

Instead, what if you switched it around and had the keys in this array be letters, and the values be the corresponding score. So it would look like

```ruby
LETTERS = {
  'a': 1,
  'b': 3,
  # ...
}
```

Then looking up a letter would be as simple as `LETTERS[letter]`.

File: specs/scrabble_scoring_spec.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+require 'minitest/autorun'
+require 'minitest/reporters'
+require 'minitest/skip_dsl'
+require_relative '../lib/scrabble_scoring'
+Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new
+
+describe "Scrabble::Scoring" do
+  it "creates a scrabble scoring class" do
+    game = Scrabble::Scoring.new
+    game.must_be_kind_of Scrabble::Scoring
```
Comment (droberts-sea) [inline specs/scrabble_scoring_spec.rb:12]: I'm not sure if this test is necessary. Since we're not supposed to instantiate `Scrabble::Scoring`, this tests a behavior that we won't ever take advantage of.

File: specs/scrabble_player_spec.rb [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  end
+end
+
+describe "Player#won?" do
+  it "Returns true, if the player has over 100 points" do
+    player = Scrabble::Player.new("Mickey")
+    player.play("qqqqq")
+    player.play("zzzzza")
+    player.won?.must_equal true
+  end
```
Comment (droberts-sea) [inline specs/scrabble_player_spec.rb:70]: I would like to see some more edge case testing here! What happens at exactly 99 points and at 100? Does it flip over as expected?

File: lib/scrabble_tilebag.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
+require_relative './scrabble_scoring'
+# require_relative './scrabble_player'
+
+module Scrabble
+  class TileBag
+    attr_reader :all_tiles
+    attr_accessor :picked_tiles
+    def initialize
+      @all_tiles = {
```
Comment (droberts-sea) [inline lib/scrabble_tilebag.rb:9]: I like the idea of this data representation - it's very easily human readable.

However, storing it in a hash like this makes your `draw_tiles` method very complex.

A cleaner strategy might be to keep the default tiles outside the `initialize` method as a constant, and make a copy of it inside `initialize`. Something like

```ruby
class TileBag
  DEFAULT_TILES = { A: 9, B: 2, ... }
  def initialize
    @tiles = []
    DEFAULT_TILES.each do |letter, frequency|
      @tiles += [letter] * frequency
    end
  end
end
```

Then `@tiles` is a big array of all the tiles in the bag, which should make drawing from it much easier.

File: lib/scrabble_tilebag.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    end
+
+    def draw_tiles(num)
+      num.times do
+        picked_tile = @all_tiles.keys.sample
+        @picked_tiles << picked_tile
+        @all_tiles[picked_tile] -= 1
+        if @all_tiles[picked_tile] == 0
+          @all_tiles.delete(picked_tile)
+        end
```
Comment (droberts-sea) [inline lib/scrabble_tilebag.rb:26]: This works, but it is initially as likely to produce a 'z' as an 'a', which isn't what we'd expect from a real-life bag full of scrabble tiles.

File: lib/scrabble_tilebag.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      end
+      return @picked_tiles
+    end
+
+    def tiles_remaining
+      @all_tiles.values.inject(:+)
```
Comment (droberts-sea) [inline lib/scrabble_tilebag.rb:32]: Note: `sum` is equivalent to `inject(:+)`, an a fair bit cleaner

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

|  Feature 	|   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions 	|   yes	|
|   Both Teammates contributed to the codebase	|   N/A	|
|   Regular Commits with meaningful commit messages 	|   yes	|
|  **Scoring class**	|   	|
|  `score` method	|   yes	|
|   Uses appropriate data structure to store the letter score	|  see inline comment 	|
|   Appropriately handles edge cases	|   Handles something that's not a string, but doesn't handle bogus strings (empty, or something like `"$%^&"`)	|
|   Tests `score` edge cases (one-letter word, 7 letter word)	|   no	|
|  `highest_scoring_word` method	|   yes	|
|   Appropriately handles edge cases	|   yes	|
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	|   Would like to see a little more here. What happens for an empty array? What about an array with bogus words?	|
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|   yes	|
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|   yes	|
|  Returns `true` or `false` from `won?` method |  `true` or `nil`, which is close enough |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |   yes  |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|   see inline comment	|
|  `draw_tiles` method uses the parameter to draw a set of random tiles |   yes	|
| `tiles_remaining` method returns a count	|   yes	|
|  Tests `draw_tiles` edge cases for different parameter inputs	|   You're still missing some cases. What happens when you draw the last tile? What happens if you've drawn all the tiles and attempt to draw some more? What happens if you attempt to draw -7 tiles?	|

I would like to start out by acknowledging that you did 2 people's worth of project on your own, which is a ton of work!

Overall this looks pretty good. Code is clean and readable, and git hygiene looks good.  There are two main areas I think you should focus going forward:

1. Data representation. I want you to be more deliberate in your choice of data structure to solve a given problem. Interesting questions include:
  - Should this be an array or a hash? Why?
  - If it's a hash, what are the keys? What are the values?
  - If it's an array, will having it in a certain order make things easier to work with?
2. Edge and failure cases. How do your methods behave when given bogus data? What about at the boundaries between bogus and good (0 vs 1 characters in a word, for example)?

File: lib/scrabble_scoring.rb (hunk 1 above)
Comment (ricecakemonster) [reply lib/scrabble_scoring.rb:3]: Wouldn't it take as much time to calculate since in both cases, the computer will go through each letter to find the value of the given letter anyway?  In case of LETTERS[S], it will still need to go through all the letters until it finds key "s" as well.  I don't see the benefit of using this way... Although it looks easier to read, I believe.  Is it(not using each loop) considered DRYer?

File: lib/scrabble_tilebag.rb (hunk 5 above)
Comment (ricecakemonster) [reply lib/scrabble_tilebag.rb:26]: Excellent point!  I didn't think of that aspect.

File: lib/scrabble_scoring.rb (hunk 1 above)
Comment (ricecakemonster) [reply lib/scrabble_scoring.rb:3]: Dan, I did test on 7 letter word for Scoring#score. (you said "no")   I guess it didn't seem obvious...
